plugins/storage/volume/adaptive/src/main/java/org/apache/cloudstack/storage/datastore/driver/AdaptiveDataStoreDriverImpl.java

```diff
@@ -0,0 +1,871 @@
+                result = new CopyCommandResult(null, null);
+                result.setSuccess(false);
+                result.setResult(e.toString());
+            }
+        } finally {
+            if (callback != null)
+                callback.complete(result);
+        }
+    }
+
+    @Override
+    public void copyAsync(DataObject srcData, DataObject destData, Host destHost,
+            AsyncCompletionCallback<CopyCommandResult> callback) {
+        copyAsync(srcData, destData, callback);
+    }
+
+    @Override
+    public boolean canCopy(DataObject srcData, DataObject destData) {
+        s_logger.debug("canCopy: Checking srcData [" + srcData.getUuid() + ":" + srcData.getType() + ":"
+                + srcData.getDataStore().getId() + " AND destData ["
+                + destData.getUuid() + ":" + destData.getType() + ":" + destData.getDataStore().getId() + "]");
+        try {
+            if (!isSameProvider(srcData)) { // TODO: change to generic
+                s_logger.debug("canCopy: No we can't -- the source provider is NOT the correct type for this driver!");
+                return false;
+            }
+
+            if (!isSameProvider(destData)) {
+                s_logger.debug("canCopy: No we can't -- the destination provider is NOT the correct type for this driver!");
+                return false;
+            }
+            s_logger.debug(
+                    "canCopy: Source and destination are the same so we can copy via storage endpoint, checking that the source actually exists");
+            StoragePoolVO poolVO = _storagePoolDao.findById(srcData.getDataStore().getId());
+            Map<String, String> details = _storagePoolDao.getDetails(srcData.getDataStore().getId());
+            ProviderAdapter api = getAPI(poolVO, details);
+
+            /**
+             * The storage provider generates its own names for snapshots which we store and
+             * retrieve when needed
+             */
+            ProviderAdapterContext context = newManagedVolumeContext(srcData);
+            ProviderAdapterDataObject srcDataObject = newManagedDataObject(srcData, poolVO);
+            if (srcData instanceof SnapshotObject) {
+                ProviderSnapshot snapshot = api.getSnapshot(context, srcDataObject);
+                if (snapshot == null) {
+                    return false;
+                } else {
+                    return true;
+                }
+            } else {
+                ProviderVolume vol = api.getVolume(context, srcDataObject);
+                if (vol == null) {
+                    return false;
+                } else {
+                    return true;
+                }
+            }
+        } catch (Throwable e) {
+            s_logger.warn("Problem checking if we canCopy", e);
+            return false;
+        }
+    }
+
+    @Override
+    public void resize(DataObject data, AsyncCompletionCallback<CreateCmdResult> callback) {
```

Comment:
If the VM is running, there is probably a need to notify libvirt of the change. 
https://github.com/apache/cloudstack/blob/a06f8a8763dc0d30b14e41d579ec73d40613e1d9/plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/driver/CloudStackPrimaryDataStoreDriverImpl.java#L437

Reply:
In the case of raw over fiberchannel, we did not find this to be the case.  the guest OS sees changes made to the raw volume on the storage array immediately.